### PR TITLE
Mixed dtype: arithmetics

### LINKED ISSUE #6385: Mixed dtype: comparison, arithmetics

### PR DESCRIPTION
**Pull request by asi1024**
~Merge #6481 first.~
Close #6385.

### CODE REVIEW

**Comment by hvy** [comment]
Just a reminder about some TODOs that can be addressed in this PR https://github.com/chainer/chainer/blob/master/chainerx_cc/chainerx/routines/linalg.cc#L26.

File: chainerx_cc/chainerx/routines/math.cc [hunk 1]
```diff
@@ -152,18 +168,23 @@ void SubtractImpl(const Array& x1, const Array& x2, const Array& out) {
...
+            });
         }
         bb.Finalize();
     }
 }
 
 void SubtractASImpl(const Array& x1, Scalar x2, const Array& out) {
-    // TODO(hvy): dtype conversion
-    if (x1.dtype() == Dtype::kBool) {
+    if (out.dtype() == Dtype::kBool) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/math.cc]: Why not call `CheckArithmeticDtypes`?

File: chainerx_cc/chainerx/routines/math.cc [hunk 2]
```diff
@@ -81,10 +80,25 @@ void BinaryInPlace(Impl&& impl, const Array& x1, Scalar x2) {
     impl(x1, x2, x1);
 }
 
+void CheckArithmeticDtypes(DtypeKind kind1, DtypeKind kind2, DtypeKind out_kind) {
+    if (kind1 == DtypeKind::kBool || kind2 == DtypeKind::kBool || out_kind == DtypeKind::kBool) {
+        throw DtypeError{"Boolean arguments are not supported in arithmetic functions."};
+    }
+    if (out_kind != DtypeKind::kFloat && (kind1 == DtypeKind::kFloat || kind2 == DtypeKind::kFloat)) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/math.cc]: Why is this check needed?

File: chainerx_cc/chainerx/routines/math.cc [hunk 3]
```diff
@@ -81,10 +80,25 @@ void BinaryInPlace(Impl&& impl, const Array& x1, Scalar x2) {
     impl(x1, x2, x1);
 }
 
+void CheckArithmeticDtypes(DtypeKind kind1, DtypeKind kind2, DtypeKind out_kind) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/math.cc]: Do you need `out_kind` in this check?

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 4]
```diff
@@ -16,7 +16,7 @@ def setup(self):
         if hasattr(self, 'expected_dtypes'):
             self.dtype, self.out_dtype = self.expected_dtypes
         elif not hasattr(self, 'out_dtype'):
-            self.out_dtype = None
+            self.out_dtype = self.dtype
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: I feel these 4 lines are difficult to understand...

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 5]
```diff
@@ -119,6 +138,18 @@ def _convert_numpy_scalar(scalar, dtype):
     return numpy.dtype(dtype).type(scalar)
 
 
+class MathScalarTestBase(object):
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Why is this needed?
I feel it's too complicated.

File: chainerx_cc/chainerx/routines/math.cc (hunk 3 above)
Comment (niboshi) [inline chainerx_cc/chainerx/routines/math.cc]: Write some comment about this function, especially the fact that `Multiply` is an exception.

File: chainerx_cc/chainerx/routines/math.cc (hunk 2 above)
Comment (asi1024) [reply chainerx_cc/chainerx/routines/math.cc]: It is because `x1 += x2` (where `x1` is of integer dtype and `x2` is of float dtype) should raise a `DtypeError`.

**Comment by niboshi** [comment]
#6635 is the blocker.

**Comment by asi1024** [comment]
Rebased.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
cpplint:
```
/home/travis/build/chainer/chainer/chainerx_cc/chainerx/routines/math.cc:350:  Add #include <utility> for move  [build/include_what_you_use] [4]
```

**Comment by niboshi** [comment]
```
___________________________ test_view[shape0-bool_] ____________________________
shape = (), dtype = 'bool_'
    def test_view(shape, dtype):
        array = array_utils.create_dummy_ndarray(chainerx, shape, dtype)
        view = array.view()
    
        chainerx.testing.assert_array_equal_ex(view, array)
        assert view.device is chainerx.get_default_device()
    
        # inplace modification
        if array.size > 0:
>           array += array
E           chainerx.DtypeError: Boolean arguments are not allowed in arithmetic functions.
tests/chainerx_tests/unit_tests/test_array.py:184: DtypeError
```

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins has passed.
Ignoring Travis CI because it takes too much time in the queue.
